src/esp/sensor/FisheyeSensor.h

```diff
@@ -0,0 +1,166 @@
+   * plane.
+   * In practice, fx and fy can differ for a number of reasons. See
+   * details here: http://ksimek.github.io/2013/08/13/intrinsic/
+   */
+  Magnum::Vector2 focalLength;
+  /**
+   * @brief Principal Point Offset in pixel, cx, cy, location of the principal
+   * point relative to the image plane's origin.
+   */
+  Magnum::Vector2 principalPointOffset;
+
+  /**
+   * @brief Constructor
+   */
+  FisheyeSensorSpec();
+
+  /**
+   * @brief operator ==, check if 2 specs are equal
+   */
+  bool operator==(const FisheyeSensorSpec& a) const;
+
+  /**
+   * @brief check if the specification is legal
+   */
+  void sanityCheck() override;
+  ESP_SMART_POINTERS(FisheyeSensorSpec)
+};
+
+struct FisheyeSensorDoubleSphereSpec : public FisheyeSensorSpec {
+  /**
+   * @brief alpha and xi are specific to "double sphere" camera model.
+   * see details (value ranges) in:
+   * Vladyslav Usenko, Nikolaus Demmel and Daniel Cremers: The Double Sphere
+   * Camera Model, The International Conference on 3D Vision (3DV), 2018
+   */
+  float alpha = 0.59;
+  float xi = -0.18;
+  /**
+   * @brief constructor
+   */
+  FisheyeSensorDoubleSphereSpec() : FisheyeSensorSpec() {}
+  /**
+   * @brief check if the specification is legal
+   */
+  void sanityCheck() override;
+  ESP_SMART_POINTERS(FisheyeSensorDoubleSphereSpec)
+};
+
+// TODO:
+// struct FisheyeFieldOfViewSpec : public FisheyeSensorSpec {};
+// struct FisheyeKannalaBrandtSpec : public FisheyeSensorSpec {};
+
+class FisheyeSensor : public VisualSensor {
+ public:
+  /**
+   * @brief constructor
+   * NOTE: the status of the camera sensor is "valid" after construction, and
+   * user can use them immediately
+   */
+  explicit FisheyeSensor(scene::SceneNode& cameraNode,
+                         const FisheyeSensorSpec::ptr& spec);
+  /**
+   * @brief destructor
+   */
+  ~FisheyeSensor() override = default;
+  /**
+   * @brief Draw an observation to the frame buffer
+   * @return true if success, otherwise false (e.g., frame buffer is not set)
+   * @param[in] sim Instance of Simulator class for which the observation needs
+   *                to be drawn
+   */
+  bool drawObservation(sim::Simulator& sim) override;
+
+  /**
+   * @brief Returns the parameters needed to unproject depth for the sensor.
+   *
+   * Will always be @ref Corrade::Containers::NullOpt for the base sensor
+   * class as it has no projection parameters
+   */
+  Corrade::Containers::Optional<Magnum::Vector2> depthUnprojection()
+      const override;
+
+  static constexpr const char* FISH_EYE_SHADER_KEY_TEMPLATE =
+      "fisheye-model-type={}-flags={}";
+
+  gfx::RenderCamera* getRenderCamera() = delete;
+
+ protected:
+  FisheyeSensorSpec::ptr fisheyeSensorSpec_ =
+      std::dynamic_pointer_cast<FisheyeSensorSpec>(spec_);
+  // raw pointer only, we can create it but let magnum to handle the memory
+  // recycling when releasing it.
+  gfx::CubeMapCamera* cubeMapCamera_;
```

Comment:
Does this mean, if I create and delete lots of FisheyeSensor instances, the cubeMapCamera_ instances will hang around unless/until the related scene nodes are deleted?

Reply:
Yes and no. :)
Yes, that is correct. `cubeMapCamera_` is a feature attaching to a node and will be destroyed when the node is gone. If the node is still there, this feature will be there.

No. it will not happen after this [PR](https://github.com/facebookresearch/habitat-sim/pull/1118). If you would like to delete a sensor, you actually delete the node it directly attaches to. We enforce it. And such a "sensor node" will be a "leaf node" when created, which means, when you delete it, you do not need to worry if there is a subtree being connected to it.